### DISCUSSION TITLE
A monorepo for testing local changes to zigbee2mqtt and herdsman

### DISCUSSION THREAD

**Discussion by deviantintegral** [Koenkk/zigbee2mqtt#25105] [Show and tell]
Over the past few years, I've often found myself wanting to build changes I've made to zigbee2mqtt into a docker container I can run with my real home network. I had some basic hacks that worked with `npm` and `npm link`, but those all broke with the switch to pnpm.

Since the project is using pnpm now, I've created a repository that configures each of them inside a pnpm workspace. This lets you:

1. Run arbitrary versions of zigbee-herdsman, zigbee-herdsman-converters, and zigbee2mqtt-frontend with zigbee2mqtt. Of course, there's no guarantee that the combination you choose will actually work.
2. Apply patches from PRs (or your own code) to any of the 4 projects.
3. Run the entire build inside docker (instead of on the host and copying in, as is done today for the production images).
4. Cross-compile from faster machines (or alternate architectures like an Apple Silicon Mac) and push the images to another host.

https://github.com/deviantintegral/zigbee2mqtt-dev-repo

Unfortunately, I wasn't able to make this work with the latest stable version of zigbee2mqtt. There's code in zigbee2mqtt that interacts with package.json, and I assume that was changed in the dev branch. I don't think it's worth fixing at this point - if this repo is of interest to you, you should probably be testing the upcoming 2.0 release anyways!

This does have me thinking that perhaps converting to a monorepo for these 4 packages is a good solution. Certainly the frontend is coupled to zigbee2mqtt pretty tightly, given it's an external project that is then consumed back into zigbee2mqtt. But I'd love to hear from others what sort of workflow they're using if this one isn't the best!

**Comment by Koenkk** [comment]
Looks good! Currently I'm using `pnpm link` which works fine for me.

**Reply by deviantintegral** [reply]
It's funny, I started with `pnpm link`. I was using this as an opportunity to try out some new AI-assisted dev tools, and at one point it made a pretty good case for why I should use a workspace instead when I was debugging issues with the local version not being used. I bet now that I know more about pnpm I could go back and make it work with link, but 🤷 I'm unblocked on local testing so that's good enough for now!

**Comment by deviantintegral** [comment]
> Cross-compile from faster machines (or alternate architectures like an Apple Silicon Mac)

I noticed that vite seems to hang in Rosetta compiling the frontend. I've run into issues with qemu before, but not rosetta. I ended up just pushing to an x64 host instead, but if anyone else runs into it and solves it I'd love to hear how!
